diff -ruN clean/kernel/supercalls.c work/kernel/supercalls.c
--- a/kernel/supercalls.c	2026-03-02 18:36:28.895068295 +0100
+++ b/kernel/supercalls.c	2026-03-03 19:26:43.515530216 +0100
@@ -345,7 +345,9 @@
     ret = ksu_set_app_profile(&cmd.profile);
     if (!ret) {
         ksu_persistent_allow_list();
+#ifndef CONFIG_KSU_SUSFS
         ksu_mark_running_process();
+#endif
     }
     return ret;
 }
@@ -492,12 +494,12 @@
 
     ret = strncpy_from_user(mnt, cmd.arg, sizeof(mnt));
     if (ret < 0) {
-        pr_err("nuke ext4 copy mnt failed: %ld\\n", ret);
-        return -EFAULT; // 或者 return ret;
+        pr_err("nuke ext4 copy mnt failed: %ld\n", ret);
+        return -EFAULT;
     }
 
     if (ret == sizeof(mnt)) {
-        pr_err("nuke ext4 mnt path too long\\n");
+        pr_err("nuke ext4 mnt path too long\n");
         return -ENAMETOOLONG;
     }
 
